lib/checks/lists/only-listitems.js

```diff
@@ -1,18 +1,84 @@
-var bad = [],
-	permitted = ['STYLE', 'META', 'LINK', 'MAP', 'AREA', 'SCRIPT', 'DATALIST', 'TEMPLATE'],
-	hasNonEmptyTextNode = false;
-
-virtualNode.children.forEach(({ actualNode }) => {
-	var nodeName = actualNode.nodeName.toUpperCase();
-	if (actualNode.nodeType === 1 && nodeName !== 'LI' && permitted.indexOf(nodeName) === -1) {
-		bad.push(actualNode);
-	} else if (actualNode.nodeType === 3 && actualNode.nodeValue.trim() !== '') {
-		hasNonEmptyTextNode = true;
-	}
-});
-
-if (bad.length) {
-	this.relatedNodes(bad);
+const ALLOWED_TAGS = [
+	'style',
+	'meta',
+	'link',
+	'map',
+	'area',
+	'script',
+	'datalist',
+	'template'
+];
+
+const getIsListItemRole = (r, t) => {
+	let out = false;
+	out = (r === 'listitem' || (t === 'li' && !r));
```

Comment:
`!r` should test that this is a valid non-abstract role, not just that it's null. This entire function could do with some cleanup. Don't use single letter variables, and just return on this line rather than assign it to a var and immediately return that.